File: android/libapp/src/main/java/com/topdon/ble/BluetoothManager.java
package com.topdon.ble;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.topdon.ble.callback.MtuChangeCallback;
import com.topdon.ble.callback.NotificationChangeCallback;
import com.topdon.ble.callback.ReadCharacteristicCallback;
import com.topdon.commons.UUIDManager;
import com.topdon.commons.observer.Observable;
import com.topdon.commons.observer.Observe;
import com.topdon.commons.poster.RunOn;
import com.topdon.commons.poster.Tag;
import com.topdon.commons.poster.ThreadMode;
import com.topdon.commons.util.LLog;
import com.topdon.commons.util.StringUtils;
import com.topdon.lms.sdk.xutils.common.util.MD5;

import org.greenrobot.eventbus.EventBus;

import java.util.UUID;
import java.util.concurrent.LinkedBlockingQueue;

@SuppressLint("MissingPermission")
public class BluetoothManager implements EventObserver {
    public static boolean iSReset = false;
    public static boolean isSending = false;
    public static boolean isClickStopCharging = false;
    private static BluetoothManager instance = null;
    private Device mDevice;
    private Connection connection;
    public static boolean isReceiveBleData = false;
    private BluetoothGattCharacteristic writeCharact = null;

    public static BluetoothManager getInstance() {
        if (instance == null)
            instance = new BluetoothManager();
        return instance;
    }

    public BluetoothManager() {
    }

    public Device getDevice() {
        return mDevice;
    }

    private void setMTUValue() {
        if (mDevice.isConnected()) {

            Log.e("bcf_ble", "连接设备名称：" + mDevice.getName() + "");
            RequestBuilder<MtuChangeCallback> builder = null;
            if (mDevice.getName().contains("T-darts") || mDevice.getName().contains("TD")) {
                builder = new RequestBuilderFactory().getChangeMtuBuilder(240);
            } else {
                builder = new RequestBuilderFactory().getChangeMtuBuilder(503);
            }
            Request request = builder.setCallback(new MtuChangeCallback() {
                @Override
                public void onMtuChanged(@NonNull Request request, int mtu) {
                    Log.d("wangchen", "MTU修改成功，新值：" + mtu);
                    setReadCallback();
                }

                @Override
                public void onRequestFailed(@NonNull Request request, int failType, @Nullable Object value) {
                    Log.d("bcf", "MTU修改失败");
                }

            }).build();
            connection.execute(request);
        }
    }

    private void setReadCallback() {
        if (mDevice.isConnected()) {
            isSending = false;

            boolean isEnabled = connection.isNotificationOrIndicationEnabled(UUID.fromString(UUIDManager.SERVICE_UUID), UUID.fromString(UUIDManager.NOTIFY_UUID));
            LLog.w("bcf_ble", "是否打开了Notifycation: " + isEnabled);
            RequestBuilder<NotificationChangeCallback> builder = new RequestBuilderFactory().getSetNotificationBuilder(UUID.fromString(UUIDManager.SERVICE_UUID), UUID.fromString(UUIDManager.NOTIFY_UUID), true);
            RequestBuilder<ReadCharacteristicCallback> builder1 = new RequestBuilderFactory().getReadCharacteristicBuilder(UUID.fromString(UUIDManager.SERVICE_UUID), UUID.fromString(UUIDManager.READ_UUID));

            builder.build().execute(connection);
            builder1.build().execute(connection);
        }
    }

    public void setCancelListening() {
        Observable observable = EasyBLE.getInstance().getObservable();
        if (observable != null) {
            EasyBLE.getInstance().unregisterObserver(this);
        }
    }

    public Connection connect(Device device) {
        mDevice = device;
        ConnectionConfiguration config = new ConnectionConfiguration();
        config.setConnectTimeoutMillis(10000);
        config.setRequestTimeoutMillis(7000);
        config.setAutoReconnect(false);
        config.setReconnectImmediatelyMaxTimes(3);
        connection = EasyBLE.getInstance().connect(device, config, this);
        connection.setBluetoothGattCallback(new BluetoothGattCallback() {
            @Override
            public void onCharacteristicWrite(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
                Log.d("ble_bcf_data", "原始写入数据状态：status: " + status + "  内容：" + StringUtils.toHex(characteristic.getValue()));
                setBleData("原始写入数据状态：status: " + status + "  内容：" + StringUtils.toHex(characteristic.getValue()));
            }
        });
        return connection;
    }

    public Connection connect(String mac,String name){
        ConnectionConfiguration configuration = new ConnectionConfiguration();
        configuration.setConnectTimeoutMillis(10000);
        configuration.setRequestTimeoutMillis(7000);
        configuration.setAutoReconnect(false);
        configuration.setReconnectImmediatelyMaxTimes(3);
        connection = EasyBLE.getInstance().connect(mac, configuration, this);
        mDevice = connection.getDevice();
        mDevice.setName(name);
        return connection;
    }

    public void release() {
        Log.d("bcf", "释放所有BLE连接");
        EasyBLE.getInstance().disconnectConnection(mDevice);
        EasyBLE.getInstance().release();
        EasyBLE.getInstance().releaseConnection(mDevice);
    }

    public boolean isConnected() {
        if (mDevice == null)
            return false;
        return mDevice.isConnected();
    }

    @Tag("onConnectionStateChanged")
    @Observe
    @RunOn(ThreadMode.MAIN)
    @Override
    public void onConnectionStateChanged(@NonNull Device device) {
        if (device.getConnectionState() != ConnectionState.SERVICE_DISCOVERED || device.getConnectionState() != ConnectionState.DISCONNECTED) {
            EventBus.getDefault().post(device.getConnectionState());
            Log.e("wangchen", "发送广播--" + device.getConnectionState());
        }
        Log.d("ywq", "MyObserver 连接状态：" + device.getConnectionState() + " 是否已连接： " + device.isConnected() + "-----名称：" + device.getName() + "-------mac: " + device.getAddress());
        switch (device.getConnectionState()) {
            case SCANNING_FOR_RECONNECTION:
                break;
            case CONNECTING:
                break;
            case CONNECTED:
                break;
            case DISCONNECTED:
                EventBus.getDefault().post(ConnectionState.DISCONNECTED.name());
                break;
            case RELEASED:
                EventBus.getDefault().post(ConnectionState.RELEASED.name());
                break;
            case SERVICE_DISCOVERED:
                setMTUValue();

                if (device.isConnected()) {
                    EventBus.getDefault().post(ConnectionState.SERVICE_DISCOVERED.name());
                }
                break;
        }
    }

    @Override
    public void onConnectFailed(Device device, int failType) {
        Log.e("bcf_ble", "连接失败" + device.getName());
        EventBus.getDefault().post(device.getConnectionState());
    }

    @Override
    public void onConnectTimeout(Device device, int type) {
        Log.e("bcf_ble", "连接超时");
    }

    @Observe
    @Override
    public void onNotificationChanged(@NonNull Request request, boolean isEnabled) {
        String typeTag = "";
        if (request.getType() == RequestType.SET_NOTIFICATION) {
            typeTag = "通知";
            EventBus.getDefault().post(ConnectionState.MTU_SUCCESS);
        } else {
            typeTag = "Indication";
        }
        Log.d("bcf_ble", "onNotificationChanged ：" + typeTag + "：" + (isEnabled ? "开启" : "关闭"));
    }

    public boolean writeBuletoothData(byte[] data) {
        if (mDevice == null || !mDevice.isConnected()) {
            return false;
        }
        writeCharact = connection.getCharacteristic(UUID.fromString(UUIDManager.SERVICE_UUID), UUID.fromString(UUIDManager.WRITE_UUID));
        connection.getGatt().setCharacteristicNotification(writeCharact, true);

        writeCharact.setWriteType(BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE);
        writeCharact.setValue(data);

        return connection.getGatt().writeCharacteristic(writeCharact);
    }

    @Observe
    @Override
    public void onCharacteristicRead(Request request, byte[] value) {

        String data = StringUtils.toHex(value);

    }

    @Observe
    @Override
    public void onCharacteristicChanged(Device device, UUID service, UUID characteristic, byte[] value) {
        Log.e("ble_bcf_data", "接收蓝牙数据：" + StringUtils.toHex(value));
        EventBus.getDefault().post(value);
    }

    public static void setBleData(String message) {

    }

}

File: android/libapp/src/main/java/com/topdon/ble/BluetoothManager.kt
package com.topdon.ble

import android.annotation.SuppressLint
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.util.Log
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.topdon.ble.callback.MtuChangeCallback
import com.topdon.ble.callback.NotificationChangeCallback
import com.topdon.ble.callback.ReadCharacteristicCallback
import com.topdon.commons.UUIDManager
import com.topdon.commons.observer.Observable
import com.topdon.commons.observer.Observe
import com.topdon.commons.poster.RunOn
import com.topdon.commons.poster.Tag
import com.topdon.commons.poster.ThreadMode
import com.topdon.commons.util.LLog
import com.topdon.commons.util.StringUtils
import org.greenrobot.eventbus.EventBus
import java.util.UUID

@SuppressLint("MissingPermission")
class BluetoothManager private constructor() : EventObserver {
    
    private var mDevice: Device? = null
    private var connection: Connection? = null
    private var writeCharact: BluetoothGattCharacteristic? = null

    companion object {
        @JvmStatic
        var iSReset = false
        
        @JvmStatic
        var isSending = false
        
        @JvmStatic
        var isClickStopCharging = false
        
        @JvmStatic
        var isReceiveBleData = false
        
        @Volatile
        private var instance: BluetoothManager? = null

        @JvmStatic
        fun getInstance(): BluetoothManager {
            return instance ?: synchronized(this) {
                instance ?: BluetoothManager().also { instance = it }
            }
        }

        @JvmStatic
        fun setBleData(message: String) {
            // Implementation for setBleData if needed
        }
    }

    fun getDevice(): Device? = mDevice

    private fun setMTUValue() {
        mDevice?.let { device ->
            if (device.isConnected) {
                Log.e("bcf_ble", "连接设备名称：${device.name}")
                
                val mtuValue = when {
                    device.name?.contains("T-darts") == true || 
                    device.name?.contains("TD") == true -> 240
                    else -> 503
                }
                
                val builder = RequestBuilderFactory().getChangeMtuBuilder(mtuValue)
                val request = builder.setCallback(object : MtuChangeCallback {
                    override fun onMtuChanged(request: Request, mtu: Int) {
                        Log.d("wangchen", "MTU修改成功，新值：$mtu")
                        setReadCallback()
                    }

                    override fun onRequestFailed(request: Request, failType: Int, value: Any?) {
                        Log.d("bcf", "MTU修改失败")
                    }
                }).build()
                
                connection?.execute(request)
            }
        }
    }

    private fun setReadCallback() {
        mDevice?.let { device ->
            if (device.isConnected) {
                isSending = false

                val serviceUUID = UUID.fromString(UUIDManager.SERVICE_UUID)
                val notifyUUID = UUID.fromString(UUIDManager.NOTIFY_UUID)
                val readUUID = UUID.fromString(UUIDManager.READ_UUID)

                val isEnabled = connection?.isNotificationOrIndicationEnabled(serviceUUID, notifyUUID) ?: false
                LLog.w("bcf_ble", "是否打开了Notifycation: $isEnabled")

                val notificationBuilder = RequestBuilderFactory()
                    .getSetNotificationBuilder(serviceUUID, notifyUUID, true)
                val readBuilder = RequestBuilderFactory()
                    .getReadCharacteristicBuilder(serviceUUID, readUUID)

                connection?.let { conn ->
                    notificationBuilder.build().execute(conn)
                    readBuilder.build().execute(conn)
                }
            }
        }
    }

    fun setCancelListening() {
        val observable = EasyBLE.getInstance().observable
        observable?.let {
            EasyBLE.getInstance().unregisterObserver(this)
        }
    }

    fun connect(device: Device): Connection? {
        mDevice = device
        val config = ConnectionConfiguration().apply {
            connectTimeoutMillis = 10000
            requestTimeoutMillis = 7000
            isAutoReconnect = false
            reconnectImmediatelyMaxTimes = 3
        }
        
        connection = EasyBLE.getInstance().connect(device, config, this)
        connection?.setBluetoothGattCallback(object : BluetoothGattCallback() {
            override fun onCharacteristicWrite(
                gatt: BluetoothGatt,
                characteristic: BluetoothGattCharacteristic,
                status: Int
            ) {
                val message = "原始写入数据状态：status: $status  内容：${StringUtils.toHex(characteristic.value)}"
                Log.d("ble_bcf_data", message)
                setBleData(message)
            }
        })
        return connection
    }

    fun connect(mac: String, name: String): Connection? {
        val configuration = ConnectionConfiguration().apply {
            connectTimeoutMillis = 10000
            requestTimeoutMillis = 7000
            isAutoReconnect = false
            reconnectImmediatelyMaxTimes = 3
        }
        
        connection = EasyBLE.getInstance().connect(mac, configuration, this)
        mDevice = connection?.device?.apply { 
            this.name = name 
        }
        return connection
    }

    fun release() {
        Log.d("bcf", "释放所有BLE连接")
        mDevice?.let { device ->
            EasyBLE.getInstance().apply {
                disconnectConnection(device)
                release()
                releaseConnection(device)
            }
        }
    }

    fun isConnected(): Boolean = mDevice?.isConnected == true

    @Tag("onConnectionStateChanged")
    @Observe
    @RunOn(ThreadMode.MAIN)
    override fun onConnectionStateChanged(device: Device) {
        val state = device.connectionState
        if (state != ConnectionState.SERVICE_DISCOVERED && state != ConnectionState.DISCONNECTED) {
            EventBus.getDefault().post(state)
            Log.e("wangchen", "发送广播--$state")
        }
        
        Log.d("ywq", "MyObserver 连接状态：$state 是否已连接：${device.isConnected}-----名称：${device.name}-------mac: ${device.address}")
        
        when (state) {
            ConnectionState.SCANNING_FOR_RECONNECTION -> {}
            ConnectionState.CONNECTING -> {}
            ConnectionState.CONNECTED -> {}
            ConnectionState.DISCONNECTED -> {
                EventBus.getDefault().post(ConnectionState.DISCONNECTED.name)
            }
            ConnectionState.RELEASED -> {
                EventBus.getDefault().post(ConnectionState.RELEASED.name)
            }
            ConnectionState.SERVICE_DISCOVERED -> {
                setMTUValue()
                if (device.isConnected) {
                    EventBus.getDefault().post(ConnectionState.SERVICE_DISCOVERED.name)
                }
            }
        }
    }

    override fun onConnectFailed(device: Device, failType: Int) {
        Log.e("bcf_ble", "连接失败${device.name}")
        EventBus.getDefault().post(device.connectionState)
    }

    override fun onConnectTimeout(device: Device, type: Int) {
        Log.e("bcf_ble", "连接超时")
    }

    @Observe
    override fun onNotificationChanged(request: Request, isEnabled: Boolean) {
        val typeTag = if (request.type == RequestType.SET_NOTIFICATION) {
            EventBus.getDefault().post(ConnectionState.MTU_SUCCESS)
            "通知"
        } else {
            "Indication"
        }
        Log.d("bcf_ble", "onNotificationChanged ：$typeTag：${if (isEnabled) "开启" else "关闭"}")
    }

    fun writeBuletoothData(data: ByteArray): Boolean {
        val device = mDevice
        val conn = connection
        
        if (device == null || !device.isConnected || conn == null) {
            return false
        }

        val serviceUUID = UUID.fromString(UUIDManager.SERVICE_UUID)
        val writeUUID = UUID.fromString(UUIDManager.WRITE_UUID)
        
        writeCharact = conn.getCharacteristic(serviceUUID, writeUUID)?.apply {
            writeType = BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
            value = data
        }

        writeCharact?.let { characteristic ->
            conn.gatt?.setCharacteristicNotification(characteristic, true)
            return conn.gatt?.writeCharacteristic(characteristic) ?: false
        }
        
        return false
    }

    @Observe
    override fun onCharacteristicRead(request: Request, value: ByteArray) {
        val data = StringUtils.toHex(value)
        // Handle characteristic read if needed
    }

    @Observe
    override fun onCharacteristicChanged(device: Device, service: UUID, characteristic: UUID, value: ByteArray) {
        Log.e("ble_bcf_data", "接收蓝牙数据：${StringUtils.toHex(value)}")
        EventBus.getDefault().post(value)
    }
}